src/shims/native_lib/mod.rs

```diff
@@ -267,6 +270,189 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
 
         interp_ok(())
     }
+
+    /// Extract the value from the result of reading an operand from the machine
+    /// and convert it to a `OwnedArg`.
+    fn op_to_ffi_arg(&self, v: &OpTy<'tcx>, tracing: bool) -> InterpResult<'tcx, OwnedArg> {
+        let this = self.eval_context_ref();
+
+        // This should go first so that we emit unsupported before doing a bunch
+        // of extra work for types that aren't supported yet.
+        let ty = this.ty_to_ffitype(v.layout.ty)?;
+
+        // Now grab the bytes of the argument.
+        let bytes = match v.as_mplace_or_imm() {
+            either::Either::Left(mplace) => {
+                // Get the alloc id corresponding to this mplace, alongside
+                // a pointer that's offset to point to this particular
+                // mplace (not one at the base addr of the allocation).
+                let mplace_ptr = mplace.ptr();
+                let sz = mplace.layout.size.bytes_usize();
+                if sz == 0 {
+                    throw_unsup_format!("Attempting to pass a ZST over FFI");
+                }
+                let (id, ofs, _) = this.ptr_get_alloc_id(mplace_ptr, sz.try_into().unwrap())?;
+                let ofs = ofs.bytes_usize();
+                // Expose all provenances in the allocation within the byte
+                // range of the struct, if any.
+                let alloc = this.get_alloc_raw(id)?;
+                let alloc_ptr = this.get_alloc_bytes_unchecked_raw(id)?;
+                for prov in alloc.provenance().get_range(this, (ofs..ofs.strict_add(sz)).into()) {
+                    this.expose_provenance(prov)?;
+                }
+                // SAFETY: We know for sure that at alloc_ptr + ofs the next layout.size
+                // bytes are part of this allocation and initialised. They might be marked
+                // as uninit in Miri, but all bytes returned by `MiriAllocBytes` are
+                // initialised.
+                unsafe {
+                    Box::from(std::slice::from_raw_parts(
+                        alloc_ptr.add(ofs),
+                        mplace.layout.size.bytes_usize(),
+                    ))
+                }
+            }
+            either::Either::Right(imm) => {
+                let (first, maybe_second) = imm.to_scalar_and_meta();
```

Comment:
`to_scalar_and_meta` is meant for wide pointers, you don't have a wide pointer here. You only need to handle `ScalarPair` since `Scalar` should be truly impossible. It seems worth a little helper that deals with one scalar so that you can call it twice.

Reply:
Sure, a `repr(C)` struct must be a scalar pair but this also needs to handle primitives, no? I can just match on the scalar directly instead of calling `to_scalar_and_meta()`, though

Reply:
> but this also needs to handle primitives, no?

Ah, is this also the codepath for regular scalars now? :+1: then.